### DISCUSSION TITLE
Metrics on Lenses do not receive applied filters

### DISCUSSION THREAD

**Discussion by kusab85** [laravel/nova-issues#5133] [Ideas & Feature Requests] [answered]
- Laravel Version: 9.42.0
- Nova Version: 4.19.4 (Silver Surfer)
- PHP Version: 8.0.11
- Database Driver & Version: SQL lite
- Operating System and Version: Windows 10 22H2
- Browser type and version: FF
- Reproduction Repository: https://github.com/kusab85/TestApp/tree/nova-lens-filters

### Description:

Metrics on Lenses do not receive applied filters.

Metrics property ``filters`` is ``null`` in `` calculate()`` method when invoked via lens.

### Detailed steps to reproduce the issue on a fresh Nova installation:

Let's have a metric:

```php

class PostsPerStatus extends Partition
{
    public function calculate(NovaRequest $request)
    {
        $statuses = Post::availableStatuses(true);

        logger(print_r([get_class($request), $this->filters], true));

        return $this
            ->count($request, Post::class, 'status')
            ->label(function ($v) use ($statuses) {
                return $statuses[$v] ?? 'UNKNOWN';
            });
    }
}

```

And add it to ``Post`` resource and ``PostsViaLens`` lens

```php  
    // ...Nova\Resources\Post.php
    public function cards(NovaRequest $request)
    {
        return [
            PostsPerStatus::make()->refreshWhenFiltersChange(),
        ];
    }
```
```php  
    // ...Nova\Lenses\PostsViaLens.php
    public function cards(NovaRequest $request)
    {
        return [
            PostsPerStatus::make()->refreshWhenFiltersChange(),
        ];
    }
```

---

On **Resource Index** page metric **DO receive** filters.

![obraz](https://user-images.githubusercontent.com/53086618/204801679-c0153da6-50c4-4fd4-b015-4af31c3aa8ab.png)

```log
[2022-11-30 12:51:06] local.DEBUG: Array
(
    [0] => Laravel\Nova\Http\Requests\MetricRequest
    [1] => Illuminate\Support\Collection Object
        ( [...] )
)
```

---

But on **Resource Lens** page metric **DO NOT receive** filters.

![obraz](https://user-images.githubusercontent.com/53086618/204804292-61ef0260-7e44-4a61-9c9f-55c43ae83daf.png)

```log
[2022-11-30 12:51:50] local.DEBUG: Array
(
    [0] => Laravel\Nova\Http\Requests\LensMetricRequest
    [1] => 
)
```

**Comment by crynobone** [comment, marked as answer]
It is much easier to match Eloquent query for Metric with Resource and automatically applying Filters tends to work out of the box. 

Lens, however, offers full customization to build the query using `Eloquent` or Fluent Query Builder and the filters query may not match the Metric query.

**Reply by kusab85** [reply]
Ok, I get it. It's not a bug. In order to retrieve lenses' filters I need set them manually.

```php 
   public function calculate(NovaRequest $request)
    {
        if (null === $this->filters and $request instanceof LensMetricRequest) {
            $this->setAvailableFilters(collect($request->lens()->availableFilters($request)));
        }

        // [...]
    }
```
